Sampling ify

EnricaBelfiore opened the pull request:
In this pull request the following methods are changed:

- `sandy.Fy.apply_qmatrix()`: the cut_hl option is inserted to cut all the decay modes of the nuclides with an half life larger than 100 years;

- `sandy.Fy.custom_perturbation()`: adapted to work with a list of ZAP numbers

A Notebook to show the IFYs sampling procedure is added.

EnricaBelfiore committed:
IFY sampling procedure

EnricaBelfiore committed:
Notebook

luca-fiorito-11 commented on sandy/fy.py:
```diff
@@ -534,9 +546,12 @@ def apply_qmatrix(self, zam, energy, decay_data, keep_fy_index=False):
         decay_data : `sandy.DecayData`
             Radioactive nuclide data for several isotopes.
+        cut_hl: `bool`, optional, default is `True`
+            cut all the decay modes of the nuclides with an half life larger
```
Typo

luca-fiorito-11 commented on sandy/fy.py:
```diff
@@ -404,11 +404,11 @@ def custom_perturbation(self, zam, mt, e, zap, pert):
-            applied.
-        pert : `float`
-            Perturbation coefficients as ratio values.
+        zap : `int` or `list`
```
I don't like `zap` and `pert` because they are redundant when multiple perturbations are given.
Then, I would rather have `pert` as a list.

Also, the scalar version can be also generalized as a special case of the vector version, with the list containing only one element.

EnricaBelfiore replied:
I agree it's redundant, so we can keep `pert` only as a `pd.Series` and remove `zap`. `pert` will be a row of the DataFrame `sandy.Samples.data` in the sampling procedure. If `pert` was a list, the `zap` should be passed in ascending order by the user. If you prefer to have `pert` as a list and two parameters, I will implement it and specify in the docstring that the two should be consistently sorted

luca-fiorito-11 replied:
I prefer the list, but it's up to you, considering consistency and maintainability.

luca-fiorito-11 replied:
> I agree it's redundant, so we can keep `pert` only as a `pd.Series` and remove `zap`. `pert` will be a row of the DataFrame `sandy.Samples.data` in the sampling procedure. If `pert` was a list, the `zap` should be passed in ascending order by the user. If you prefer to have `pert` as a list and two parameters, I will implement it and specify in the docstring that the two should be consistently sorted

As i see it, `zap` and `pert` should not be in ascending order, right?! They should just be in a consistent order.

luca-fiorito-11 replied:
Both ways are fine, but if you go for a series, then I would rather have a `dict`, which is equivalent. 
What do you think?

EnricaBelfiore replied:
My concern is that when filtering `Fy.data` with the list of ZAPs, the index of the resulting sub-DataFrame will be ordered ascendingly independently of the order of the list. This means that the perturbation coefficients should be ordered ascendingly from the one to apply to the nuclide to the lower ZAP. To me, using a data structure that keeps the information together would help. I can agree with the dictionary but the multiplication in line 437 ensures that a perturbation coefficient is multiplied by the fission yield of the corresponding ZAP thanks to the index, so I would have to cast the dictionary to a pd.Series. 

The implementation with the pd.Series is in the next commit

luca-fiorito-11 commented on sandy/fy.py:
```diff
@@ -422,10 +422,22 @@ def custom_perturbation(self, zam, mt, e, zap, pert):
-        mask = (df.ZAM == zam) & (df.ZAP == zap) & (df.MT == mt) & (df.E == e)
-        df.loc[mask, "FY"] = df[mask].squeeze().FY * pert
+        idx = df.query(f"ZAM=={zam} & ZAP=={zap} & MT=={mt} & E=={e}").index
+        fy_pert = df.loc[idx, :].set_index("ZAP").FY * pert
```
If pert is a list, then you don't need to match the index anymore for the product and this could be simplified

luca-fiorito-11 reviewed (commented):
Some comments on method `custom_perturbation`.
@EnricaBelfiore , I can already merge or wait. Let me know.

EnricaBelfiore committed:
zap kwarg removed
    def custom_perturbation(self, zam, mt, e, pert):
        """
        Apply a custom perturbation to a given fission yield.

        Parameters
        ----------
        zam : `int`
            ZAM number of the material to which perturbations are to be
            applied.
        mt : `int`
            MT reaction number of the FY to which perturbations are to be
            applied either 454 or 459.
        e : `float`
            Energy of the fissioning system to which which perturbations
            are to be applied.
        pert : `pd.Series`
            Perturbation coefficients as ratio values. `pert` index should be
            the corrisponding ZAP numbers.

        Returns
        -------
        `Fy`
            Fission yield instance with applied perturbation.

        Examples
        --------
        >>> tape = sandy.get_endf6_file("jeff_33", 'nfpy', 'all')
        >>> nfpy = Fy.from_endf6(tape)
        >>> pert = pd.Series([0.9], index=[551370])
        >>> nfpy_pert = nfpy.custom_perturbation(922350, 459, 0.0253, pert)
        >>> comp = nfpy_pert.data.query('ZAM==922350 & ZAP==551370 & MT==459 & E==0.0253').squeeze().FY
        >>> assert np.setdiff1d(nfpy_pert.data.values, nfpy.data.values) == comp

        >>> pert = pd.Series([0.9, 0.5], index=[541350, 551370])
        >>> nfpy_pert = nfpy.custom_perturbation(922350, 459, 0.0253, pert)
        >>> comp1 = nfpy_pert.data.query('ZAM==922350 & ZAP==551370 & MT==459 & E==0.0253').squeeze().FY
        >>> f1 = nfpy.data.query('ZAM==922350 & ZAP==551370 & MT==459 & E==0.0253').squeeze().FY
        >>> assert f1 * 0.5 == comp1
        >>> comp2 = nfpy_pert.data.query('ZAM==922350 & ZAP==541350 & MT==459 & E==0.0253').squeeze().FY
        >>> f2 = nfpy.data.query('ZAM==922350 & ZAP==541350 & MT==459 & E==0.0253').squeeze().FY
        >>> assert f2 * 0.9 == comp2
        """
        df = self.data.copy()
        idx = df.query(f"ZAM=={zam} & ZAP=={list(pert.index)} & MT=={mt} & E=={e}").index
        fy_pert = df.loc[idx, :].set_index("ZAP").FY * pert
        fy_pert.index = idx
        fy_pert.name = 'FY'
        df.update(fy_pert)
        return self.__class__(df)
